tests/rptest/tests/offset_retention_test.py

```diff
@@ -9,9 +9,135 @@
+        unique_versions = wait_for_num_versions(self.redpanda, 1)
+        assert ver_string(version) in unique_versions
+
+        # sanity check none of feature configs should exist
+        admin = Admin(self.redpanda)
+        config = admin.get_cluster_config()
+        assert config.keys().isdisjoint(self.feature_config_names)
+
+    def _perform_upgrade(self, initial_version, version):
+        """
+        1. verify upgrade of all cluster nodes
+        2. verify expected configs after upgrade
+        """
+        # upgrade all nodes to target version
+        self.installer.install(self.redpanda.nodes, version)
+        self.redpanda.restart_nodes(self.redpanda.nodes)
+
+        # wait until all nodes are running the same version
+        unique_versions = wait_for_num_versions(self.redpanda, 1)
+        assert ver_string(initial_version) not in unique_versions
+
+        # sanity check that all new feature configs exist
+        admin = Admin(self.redpanda)
+        config = admin.get_cluster_config()
+        assert config.keys() > self.feature_config_names
+
+        # configs should all have positive values
+        for name in self.feature_config_timing:
+            assert config[name] > 0
+
+        # after upgrade legacy support should be disabled
+        assert config[self.feature_config_legacy] == False
+
+    def _offset_removal_occurred(self, period):
+        rpk = RpkTool(self.redpanda)
+        group = "hey_group"
+
+        def offsets_exist():
+            desc = rpk.group_describe(group)
+            return len(desc.partitions) > 0
+
+        # consume from the group for twice as long as the retention period
+        # setting and verify that group offsets exist for the entire time.
+        start = time.time()
+        while time.time() - start < (period * 2):
+            rpk.produce(self.topic, "k", "v")
+            rpk.consume(self.topic, n=1, group=group)
+            wait_until(offsets_exist, timeout_sec=1, backoff_sec=1)
+            time.sleep(1)
+
+        # after one half life the offset should still exist
+        time.sleep(period / 2)
+        assert offsets_exist()
+
+        # after waiting for twice the retention period, it should be gone
+        time.sleep(period * 2 - period / 2)
+        return not offsets_exist()
+
+    @cluster(num_nodes=3, log_allow_list=RESTART_LOG_ALLOW_LIST)
+    @parametrize(initial_version=(22, 2, 9))
+    @parametrize(initial_version=(22, 3, 11))
+    def test_upgrade_from_pre_v23(self, initial_version):
+        """
+        1. test that retention feature doesn't work on legacy version
+        2. upgrade cluster and test that feature still doesn't work
+        3. enable legacy support on the cluster
+        4. test that retention works properly
+        """
+        period = 30
+
+        # in old cluster offset retention should not be active
+        self._validate_pre_upgrade(initial_version)
+        assert not self._offset_removal_occurred(period)
+
```

Comment:
q: Would it make sense to test a case where the cluster is in a partially upgraded state? Or would that really be testing the feature manager and not the offset retention code?

Reply:
yeh good point. im not sure. on the one hand we have to trust other sub-systems or it'd be impossible to write reasonable sized tests. otoh i might have used feature manager incorrectly. it probably boils down to this last point, and depends on how critical things are. this one is not critical wrt data loss etc...